pkg/awsclient/sts_assume_role.go

```diff
@@ -0,0 +1,153 @@
+package awsclient
+
+import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/sts"
+	"github.com/go-logr/logr"
+	awsv1alpha1 "github.com/openshift/aws-account-operator/api/v1alpha1"
+	"github.com/openshift/aws-account-operator/config"
+	"github.com/rkt/rkt/tests/testutils/logger"
+	"regexp"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+var (
+	defaultSleepDelay = 500 * time.Millisecond
+)
+
+const (
+	controllerName = "account"
+)
+
+func matchSubstring(roleID, role string) (bool, error) {
+	matched, err := regexp.MatchString(roleID, role)
+	return matched, err
+}
+
+// getSTSCredentials returns STS credentials for the specified account ARN
+func GetSTSCredentials(
+	reqLogger logr.Logger,
+	client Client,
+	roleArn string,
+	externalID string,
+	roleSessionName string) (*sts.AssumeRoleOutput, error) {
+	// Default duration in seconds of the session token 3600. We need to have the roles policy
+	// changed if we want it to be longer than 3600 seconds
+	var roleSessionDuration int64 = 3600
+	reqLogger.Info(fmt.Sprintf("Creating STS credentials for AWS ARN: %s", roleArn))
+	// Build input for AssumeRole
+	assumeRoleInput := sts.AssumeRoleInput{
+		DurationSeconds: &roleSessionDuration,
+		RoleArn:         &roleArn,
+		RoleSessionName: &roleSessionName,
+	}
+	if externalID != "" {
+		assumeRoleInput.ExternalId = &externalID
+	}
+
+	assumeRoleOutput := &sts.AssumeRoleOutput{}
+	var err error
+	for i := 0; i < 100; i++ {
+		time.Sleep(defaultSleepDelay)
+		assumeRoleOutput, err = client.AssumeRole(&assumeRoleInput)
+		if err == nil {
+			break
+		}
+		if i == 99 {
+			reqLogger.Info(fmt.Sprintf("Timed out while assuming role %s", roleArn))
+		}
+	}
+	if err != nil {
+		// Log AWS error
+		if aerr, ok := err.(awserr.Error); ok {
+			reqLogger.Error(aerr,
+				fmt.Sprintf(`New AWS Error while getting STS credentials,
+					AWS Error Code: %s,
+					AWS Error Message: %s`,
+					aerr.Code(),
+					aerr.Message()))
+		}
+		return &sts.AssumeRoleOutput{}, err
```

Comment:
This line can actually be removed. 
When `err != nil`, the var `assumeRoleOutput` will be nil, and the other way around.
So you can just write 
```
return assumeRoleOutput, err
```
at the end of the function and have the same behaviour.

Reply:
when I remove that the return statement the [TestGetSTSCredentials ](https://github.com/openshift/aws-account-operator/blob/5be3cb3945eb31bf35836985bae93af675664177/controllers/account/iam_test.go#L130) fails

Reply:
I see, that's because of the way the test is set up. Because the test is both returning an error and sts credentials in the [AssumeRole call](https://github.com/openshift/aws-account-operator/blob/5be3cb3945eb31bf35836985bae93af675664177/controllers/account/iam_test.go#L110C7-L110C7) and then expects the returned credentials to actually by empty. Imho, we should remove that check from the test, as i don't really see the value in that. But I also don't think it's that important, so i leave this up to you :)